File: src/Functional/Scala/Facade.scala
object FacadeExample:
  final class AuthService:
    def authenticate(user: String): String = s"auth($user)"

  final class InventoryService:
    def reserve(sku: String): String = s"reserve($sku)"

  final class BillingService:
    def charge(amount: Int): String = s"charge($amount)"

  final class CheckoutFacade(
      auth: AuthService,
      inventory: InventoryService,
      billing: BillingService
  ):
    def checkout(user: String, sku: String, amount: Int): String =
      List(
        auth.authenticate(user),
        inventory.reserve(sku),
        billing.charge(amount)
      ).mkString("checkout=", ">", "")

  def main(args: Array[String]): Unit =
    val facade = CheckoutFacade(AuthService(), InventoryService(), BillingService())
    println(facade.checkout("alice", "SKU-42", 499))

File: src/Functional/OCaml/facade.ml
let auth_service user = "auth(" ^ user ^ ")"
let inventory_service sku = "reserve(" ^ sku ^ ")"
let billing_service amount = "charge(" ^ string_of_int amount ^ ")"

let checkout_facade user sku amount =
  String.concat ">"
    [ auth_service user; inventory_service sku; billing_service amount ]

let () =
  Printf.printf "checkout=%s\n" (checkout_facade "alice" "SKU-42" 499)
